### ISSUE TITLE
Unit tests needed!

### ISSUE THREAD

**Issue by herrvigg** [qtranslate/qtranslate-xt#620]
This project is very hard to maintain, not only because the code is very messy and mostly undocumented but also because of the lack of tests. WP is a messy framework by itself for the integration and we all know they maintain a lot of legacy code (and crap). But at least in each theme/plugin a lot can be done to ensure we don't break stuff internally.

In this project there is almost nothing tested! I only found a few tests on the URL conversions but it's only a few functions without any real test framework and i don't even know if they have been maintained.

TL;DR: to make qTranslate viable we need a serious test strategy.

**Comment by boelle** [comment]
i have just installed the plugin and translated the main page in to danish

but i use the one page express theme and i can't translate (or i dont know how to do that) the top part of the page. also the contact form is not translated yet as i have to figure out how. its Contact Form 7 i use for that

happy to be a tester but its still day one for me and documentaion is hard to find

**Comment by herrvigg** [comment]
Hello,
i don't know these plugins but for integration i'd recommend you to check the website documentation from the legacy qTranslate-X. This one is still working the same way (might change in the future). 

https://qtranslatexteam.wordpress.com/integration/

By tests i needed automated tests but all users helping to validate are much welcome. Integration in wordpress is definitely a big challenge.

**Comment by boelle** [comment]
i read that link, but its not clear enough, but then again the org guy has turned in to fumes and blown with the wind

**Comment by herrvigg** [comment]
Yes this part is quite confusing, especially the JSON file. There should at least be a GUI in the options to edit this instead of having to go through all the options. It's not convenient but at least there's something. The best is probably to check the existing examples, including the main i18-config.json and understand little by little all the fields.

**Comment by boelle** [comment]
problem is that now things starts to get confusing and the avg. joe user simple uninstalls and find something else

and the plugin was interfering with my theme's functions.php and prevented me from submitting my site to google's HSTS preload

**Comment by herrvigg** [comment]
qTranslate has never been easy to use for advanced integration but that's also inherent to wordpress and any free plugin. It comes with some limitations but the users have to deal with that or buy real products with customer support. This is what the previous author was apparently attempting to do to monetize it. But this won't be my goal, i just want something that works for the core functionalities so that the community here can maintain it. Support for the main third-party plugins is also important but many things have to be refactored and possibly simplified. It's over-complicated and if it is released as a new official plugin some legacy stuff may be left out to make it viable.

**Comment by Panstube** [comment]
Thank you for the time that you  invest in this great plugin :) I test it about a week now and it working good on PansTube.com Things that dont work on my setup are very few - Yoast need to refresh page to change (this not concerning me as loong it work). Schema - take only main language english. But this are more integration problems. 
I`m not a coding gui  but if I can help in any way count me in.

**Comment by herrvigg** [comment]
Thanks!
I have now a better plan in mind for the tests part, it will be essential for the maintenance of this plugin which is currently very difficult because there's (almost) nothing tested and this is slowing down everything. I'll keep you updated.
